File: Modeling .NET/VB/Queens/Queens/Board.vb
Public Class Board
    Public Const QueenCount = 5
    Shared Rows() As Boolean = New Boolean(QueenCount - 1) {}
    Shared DiagsUp() As Boolean = New Boolean(2 * QueenCount - 2) {}
    Shared DiagsDown() As Boolean = New Boolean(2 * QueenCount - 2) {}
    Public Shared Queens() As Queen = New Queen(QueenCount - 1) {}

    Public Shared Function IsFree(ByVal Col As Byte, ByVal Row As Byte) As Boolean
        Return Not Rows(Row) And Not DiagsUp(Col + QueenCount - Row - 1) And Not DiagsDown(Col + Row)
    End Function

    Public Shared Sub MakeOccupied(ByVal Col As Byte, ByVal Row As Byte)
        Rows(Row) = True
        DiagsUp(Col + QueenCount - Row - 1) = True
        DiagsDown(Col + Row) = True
    End Sub

    Public Shared Sub MakeFree(ByVal Col As Byte, ByVal Row As Byte)
        Rows(Row) = False
        DiagsUp(Col + QueenCount - Row - 1) = False
        DiagsDown(Col + Row) = False
    End Sub

    Public Shared Sub Remember()
        Dim i As Integer
        For i = 0 To QueenCount - 1
            Console.Write("{0} ", Queens(i).Row)
        Next
        Console.WriteLine()
    End Sub
End Class

File: Modeling .NET/CSharp/Queens/Queens/Board.cs
using System;
using System.Collections.Generic;
using System.Text;

namespace Queens
{
    /// <summary>
    /// Глобальные определения, моделирующие шахматную доску
    /// </summary>
    class Board
    {
        /// <summary>
        /// Количество ферзей и размер доски
        /// </summary>
        public const int QueenCount = 4;

        /// <summary>
        /// Массив, описывающий занятость горизонталей
        /// </summary>
        static bool[] Rows = new bool[QueenCount];

        /// <summary>
        /// Массив, описывающий занятость диагоналей из левого нижнего угла в правый верхний
        /// </summary>
        static bool[] DiagsUp = new bool[2 * QueenCount - 1];

        /// <summary>
        /// Массив, описывающий занятость диагоналей из левого верхнего угла в правый нижний
        /// </summary>
        static bool[] DiagsDown = new bool[2 * QueenCount - 1];

        /// <summary>
        /// Массив сопрограмм, имитирующих расстановку ферзей
        /// </summary>
        static public Queen[] Queens = new Queen[QueenCount];

        /// <summary>
        /// Количество найденных расстановок
        /// </summary>
        static public int BoardCount = 0;

        /// <summary>
        /// Проверка, свободно ли поле шаматной доски
        /// </summary>
        /// <param name="Col">Номер вертикали</param>
        /// <param name="Row">Номер горизонтали</param>
        /// <returns>true, если поле свободно</returns>
        static public bool IsFree(byte Col, byte Row)
        {
            return !Rows[Row] && !DiagsUp[Col - Row + QueenCount - 1] && !DiagsDown[Col + Row];
        }

        /// <summary>
        /// Занять поле шахматной доски
        /// </summary>
        /// <param name="Col">Номер вертикали</param>
        /// <param name="Row">Номер горизонтали</param>
        static public void MakeOccupied(byte Col, byte Row)
        {
            Rows[Row] = true;
            DiagsUp[Col - Row + QueenCount - 1] = true;
            DiagsDown[Col + Row] = true;
        }

        /// <summary>
        /// Освободить поле шахматной доски
        /// </summary>
        /// <param name="Col">Номер вертикали</param>
        /// <param name="Row">Номер горизонтали</param>
        static public void MakeFree(byte Col, byte Row)
        {
            Rows[Row] = false;
            DiagsUp[Col - Row + QueenCount - 1] = false;
            DiagsDown[Col + Row] = false;
        }

        /// <summary>
        /// Отобразить текущую расстановку
        /// </summary>
        static public void Remember()
        {
            for (int i = 0; i < QueenCount; i++)
                Console.Write("{0} ", Queens[i].Row);
            Console.WriteLine();
            BoardCount++;
        }
    }
}
